File: 프로그래머스/lv3/77886. 110 옮기기/110 옮기기.js
function solution(s) {
    let answer = []
    
    s.forEach(str => {
        
        // '110' 찾아서 꺼내기
        let oneCnt = '';
        let newStr = [];
        let len = 0;
        for(let i = 0; i<str.length; i++){
            newStr.push(str[i]);
            len = newStr.length;
            if(newStr.length>2){
                if(newStr[len-3]=='1' && newStr[len-2]=='1' && newStr[len-1]=='0'){
                    oneCnt += '110';
                    newStr.pop();newStr.pop();newStr.pop();
                }
            }
        }
        // 110 넣기
        let lastZero = newStr.lastIndexOf('0')+1;
        newStr = newStr.join('');
        newStr = newStr.slice(0, lastZero) + oneCnt + newStr.slice(lastZero, newStr.length)
1
        answer.push(newStr)
    })
    
    return answer
}

File: 프로그래머스/lv3/77886. 110 옮기기/110 옮기기.java
import java.util.Stack;

class Solution {
    public String[] solution(String[] s) {
     // 110의 갯수를새고 110을 제외한 문자열을 만든다.
     // 앞에서부터 111이 반복되는 숫자를 찾아서 찾을 경우 그곳에 110을 넣고 하나도 없으면 뒤에 계속 붙인다.
        String[] result = new String[s.length];  
        
        for(int i=0; i<s.length; i++){
            int cnt = 0;
            
            Stack<Character> st = new Stack<>();
            // 문자열에서 110의 갯수와, 110을 제거한 문자열을 스텍에 담는다.
            for(int j=0; j<s[i].length(); j++){
                char a = s[i].charAt(j);
                if(st.size()>=2){
                    char b = st.pop();
                    char c = st.pop();
                    if(c=='1' && b=='1' && a=='0')
                        cnt++;
                    else{
                        st.push(c);
                        st.push(b);
                        st.push(a);
                    }
                }else{
                    st.push(a);
                }
            }
            StringBuilder sb = new StringBuilder();
            while(!st.isEmpty())
                sb.append(st.pop()); // 스택에 담긴 char들을 string으로 바꾼다.
            sb = new StringBuilder(sb.reverse().toString());// 110을 제외한 문자열을 구한다.
            
            int idx = sb.lastIndexOf("0")+1; // 0의 마지막 위치를 구한다.

            while(cnt!=0){
                sb.insert(idx,"110"); // 0의 마지막위치가 없으면 자동적으로 맨앞에 110이 추가된다.
                cnt--;
            }
            
            result[i] = sb.toString();
        }
        return result;
    }
}